Common/Orders/Fills/EquityFillModel.cs

```diff
@@ -379,12 +379,49 @@ public virtual OrderEvent MarketOnOpenFill(Security asset, MarketOnOpenOrder ord
             // have large gaps, in which case the currentBar.EndTime will be in the past
             // ASUR  | | |      [order]        | | | | | | |
             //  SPY  | | | | | | | | | | | | | | | | | | | |
-            var currentBar = asset.GetLastData();
             var localOrderTime = order.Time.ConvertFromUtc(asset.Exchange.TimeZone);
-            if (currentBar == null || localOrderTime >= currentBar.EndTime) return fill;
+            var endTime = DateTime.MinValue;
+
+            var subscribedTypes = GetSubscribedTypes(asset);
+
+            if (subscribedTypes.Contains(typeof(Tick)))
+            {
+                // Get the first valid (non-zero) tick of trade type from an open market
+                var trade = asset.Cache.GetAll<Tick>().FirstOrDefault(x =>
+                    x.TickType == TickType.Trade && x.Price > 0 && asset.Exchange.DateTimeIsOpen(x.Time));
+                if (trade != null)
+                {
+                    endTime = trade.EndTime;
+                    fill.FillPrice = trade.Price;
+                }
+            }
+            else if (subscribedTypes.Contains(typeof(TradeBar)))
+            {
+                var tradeBar = asset.Cache.GetData<TradeBar>();
+                if (tradeBar != null)
+                {
+                    // If the order was placed during the bar aggregation, we cannot use its open price
+                    if (tradeBar.Time < localOrderTime) return fill;
+
+                    // If the first data after the market opens is a quote bar (high resolution only),
+                    // the open of the last trade bar is not the open for the current day, but the last open in cache.
+                    // We need to verify whether the trade data is from the open market.
+                    if (tradeBar.Period < Resolution.Hour.ToTimeSpan() && !asset.Exchange.DateTimeIsOpen(tradeBar.Time))
+                    {
+                        return fill;
+                    }
+
+                    endTime = tradeBar.EndTime;
+                    fill.FillPrice = tradeBar.Open;
+                }
+            }
+
+            if (localOrderTime >= endTime) return fill;
```

Comment:
Minute data:  Submission Previous Day / Same Day 12am / 8am / 930am, first bar sets time at X - End time 931am
Hour data:     Submission Previous Day / Same Day 12am / 8am / 930am, first bar sets time at X  - End time 10am
Daily data:     Submission Previous Day / Same Day 12am / 8am / 930am, first bar sets time at X  - End time 12am tomorrow.

Reply:
I added this unit test to address this comment:
https://github.com/QuantConnect/Lean/pull/5129/commits/78ddf8af110a33aafb2d0fcf31a269d0f3561572#diff-18938dbae43240eab6eee79fdd2221594549c57b2b4937ba737188dd5197d9d5R556